src/jdk.internal.le/linux/native/lible/CLibrary.cpp

```diff
@@ -0,0 +1,206 @@
+    env->SetIntField(result, c_cflag, data.c_cflag);
+    env->SetIntField(result, c_lflag, data.c_lflag);
+    env->SetIntField(result, c_line, data.c_line);
+    jbyteArray c_ccValue = (jbyteArray) env->GetObjectField(result, c_cc);
+    env->SetByteArrayRegion(c_ccValue, 0, NCCS, (signed char *) data.c_cc);//TODO: cast?
+    env->SetIntField(result, c_ispeed, data.c_ispeed);
+    env->SetIntField(result, c_ospeed, data.c_ospeed);
+}
+
+/*
+ * Class:     jdk_internal_org_jline_terminal_impl_jna_linux_CLibraryImpl
+ * Method:    tcsetattr
+ * Signature: (IILjdk/internal/org/jline/terminal/impl/jna/linux/CLibrary/termios;)V
+ */
+JNIEXPORT void JNICALL Java_jdk_internal_org_jline_terminal_impl_jna_linux_CLibraryImpl_tcsetattr
+  (JNIEnv *env, jobject, jint fd, jint cmd, jobject input) {
+    termios data;
+
+    data.c_iflag = env->GetIntField(input, c_iflag);
+    data.c_oflag = env->GetIntField(input, c_oflag);
+    data.c_cflag = env->GetIntField(input, c_cflag);
+    data.c_lflag = env->GetIntField(input, c_lflag);
+    data.c_line = env->GetIntField(input, c_line);
+    jbyteArray c_ccValue = (jbyteArray) env->GetObjectField(input, c_cc);
+    env->GetByteArrayRegion(c_ccValue, 0, NCCS, (signed char *) data.c_cc);//TODO: cast?
+    data.c_ispeed = env->GetIntField(input, c_ispeed);
+    data.c_ospeed = env->GetIntField(input, c_ospeed);
+
+    if (tcsetattr(fd, cmd, &data) != 0) {
+        jobject exc = env->NewObject(lastErrorExceptionClass,
+                                     lastErrorExceptionConstructor,
+                                     errno);
+        env->Throw((jthrowable) exc);
+    }
+}
+
+/*
+ * Class:     jdk_internal_org_jline_terminal_impl_jna_linux_CLibraryImpl
+ * Method:    ioctl0
+ * Signature: (IILjdk/internal/org/jline/terminal/impl/jna/linux/CLibrary/winsize;)V
+ */
+JNIEXPORT void JNICALL Java_jdk_internal_org_jline_terminal_impl_jna_linux_CLibraryImpl_ioctl0
+  (JNIEnv *env, jobject, jint fd, jint cmd, jobject data) {
+    winsize ws;
+
+    ws.ws_row = env->GetIntField(data, ws_row);
+    ws.ws_col = env->GetIntField(data, ws_col);
+    ws.ws_xpixel = env->GetIntField(data, ws_xpixel);
+    ws.ws_ypixel = env->GetIntField(data, ws_ypixel);
+
+    if (ioctl(fd, cmd, &ws) != 0) {
+        jobject exc = env->NewObject(lastErrorExceptionClass,
+                                     lastErrorExceptionConstructor,
+                                     errno);
+        env->Throw((jthrowable) exc);
+        return ;
+    }
+
+    env->SetIntField(data, ws_row, ws.ws_row);
+    env->SetIntField(data, ws_col, ws.ws_col);
+    env->SetIntField(data, ws_xpixel, ws.ws_xpixel);
+    env->SetIntField(data, ws_ypixel, ws.ws_ypixel);
+}
+
+/*
+ * Class:     jdk_internal_org_jline_terminal_impl_jna_linux_CLibraryImpl
+ * Method:    isatty
+ * Signature: (I)I
+ */
+JNIEXPORT jint JNICALL Java_jdk_internal_org_jline_terminal_impl_jna_linux_CLibraryImpl_isatty
+  (JNIEnv *, jobject, jint fd) {
+    return isatty(fd);
```

Comment:
Do we care if the native `isatty()` returns zero? Or is this dealt with somewhere else?

Reply:
This method is only used to determine with the fd is attached to a terminal (returns 1) or not (return 0). The reasons why it is not attached to a terminal are not really important. The value is used here:
https://github.com/lahodaj/jdk/blob/4cf8f67e43f442a5c48cd30349740ac2cb638d6e/src/jdk.internal.le/unix/classes/jdk/internal/org/jline/terminal/impl/jna/JnaNativePty.java#L186